jupyter_accessibility/project.py

```diff
@@ -0,0 +1,428 @@
+        return repo
+
+
+class Package(Common):
+    """the base class of all packages
+
+    packages are built and available on package managers. the id `jupyterlab` is the `jupyterlab` package in `pip`"""
+
+    _EXTENSIONS = {}  # extension base classes define there own mapping ot manage
+
+    def tasks(self):
+        yield dict(
+            basename="setup",
+            name=f"pip:{self.name}",
+            actions=[do(f"{self.conda} python -m pip install {self.name}")],
+            clean=[do(f"{self.conda} python -m pip uninstall -y {self.name}")],
+        )
+
+    def __post_init__(self):
+        # propogate parent content to package
+        self.dir = self.parent.dir
+        self.conda = self.parent.conda
+
+
+PIP_BUILD_ENV = dict(
+    NODE_OPTS="--max-old-space-size=4096",
+    PIP_DISABLE_PIP_VERSION_CHECK="1",
+    # PIP_IGNORE_INSTALLED="1",
+    PIP_NO_BUILD_ISOLATION="1",
+    # PIP_NO_DEPENDENCIES="1",
+    PYTHONIOENCODING="utf-8",
+    PYTHONUNBUFFERED="1",
+)
+
+
+class Repo(Common):
+    """the base class for all repos
+
+    repos need instructions to build jupyter products, they are defined as doit tasks in each class.
+    these repos needed to be cloned, dev dependencies built.
+    """
+
+    _EXTENSIONS = {}
+    url: str = Field(description="the url for the project pointing to a git repo")
+    branch: str = Field("main", description="the branch, ref, or hash to build")
+    path: Path = Field(None, description="path the repository with a git repo")
+    head: Path = Field(None, description="the HEAD of the git repo")
+    package: Path = Field(None, description="the package.json")
+    links: Path = Field(None, description="the yarn links location")
+    pip: str = Field(None, description="the pip command prefix including conda")
+    yarn: Path = Field(None, description="yarn prefix including conda")
+    yarn_integrity: Path = Field(None, description="the location of the yarn integrity file")
+    # useful for hashing
+    setup: Path = Field(None, description="the setup.py target")  # assuming one is being used
+
+    def __post_init__(self):
+        self.name = get_name(self.url)
+        self.dir = self.parent.dir
+        self.path = self.dir / self.name
+        self.head = self.path / ".git" / "HEAD"
+        self.package = self.path / "package.json"
+        self.links = self.path / "yarn-links"
+        self.conda = self.parent.conda
+        self.yarn = f"{self.conda} yarn --link-folder {self.links.resolve()}"
+        self.pip = f"{self.conda} python -m pip"
+        self.yarn_integrity = self.path / "node_modules" / ".yarn-integrity"
+        self.setup = self.path / "setup.py"
+
+    def get_packages(self, where="packages"):
+        """yield package.json from all folders one level down"""
+        # this requires the repos are setup/cloned
+        yield from (self.path / where).rglob("*/package.json")
+
+    # cloning requires the env
+    # everything else requires cloning
```

Comment:
I am somehow confused by this statement but still need to go through some other bits

Reply:
the next step in this work is to set up the conditions for completed tasks in do it. we'll need to build out the DAG.

if someone does not have `git` then they will need a `conda` environment ahead of time. these are notes for me to explicitly express the logic for running the tasks.